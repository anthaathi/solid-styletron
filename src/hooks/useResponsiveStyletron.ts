import { StyleObject } from 'styletron-standard';
import type {
  AtRule,
  ObsoleteProperties,
  Property,
  StandardProperties,
  SvgProperties,
  VendorProperties,
} from 'csstype';
import { useStyletron } from './useStyletron';
import { StyletronTheme } from '../contexts/styletron';

export type ResponsiveStyleObject = Properties & {
  [key in string]: Properties[keyof Properties] | StyleObject;
};

export function useResponsiveStyletron(): [
  css: (input: ResponsiveStyleObject) => string,
  $theme: StyletronTheme
] {
  const [$css, $theme] = useStyletron();

  function css(input: ResponsiveStyleObject) {
    const returnObject: StyleObject = {};
    const mediaQueriesKeys = Object.keys($theme.mediaQueries || {});

    Object.keys(input).forEach((key) => {
      const inputElement = input[key];

      if (Array.isArray(inputElement)) {
        inputElement.forEach((item, index) => {
          const mediaQuery = $theme.mediaQueries?.[mediaQueriesKeys[index]];

          if (!mediaQuery) {
            console.warn('mediaQuery not found for index', index);
            return;
          }

          returnObject[mediaQuery] = returnObject[mediaQuery] || {};
          // @ts-ignore
          returnObject[mediaQuery][key] = item as never;
        });
        return;
      }

      // @ts-ignore
      returnObject[key] = inputElement;
    });

    return $css(returnObject);
  }

  return [css, $theme];
}

export interface KeyframesPercentageObject {
  [key: string]: Properties;
}

export type KeyframesObject = KeyframesPercentageObject & {
  from?: Properties;
  to?: Properties;
};

export type AnimationNameProperty =
  | Property.AnimationName
  | KeyframesObject
  | Property.AnimationName[]
  | KeyframesObject[];

export type FontFace = AtRule.FontFace<TLength>;

export type FontFamilyProperty = Property.FontFamily | FontFace;

type TLength = string | 0 | string[] | number[];

export type Properties = {
  animationName?: AnimationNameProperty;
  fontFamily?: FontFamilyProperty | FontFamilyProperty[];
  MozAnimationName?: AnimationNameProperty;
  WebkitAnimationName?: AnimationNameProperty;
  OAnimationName?: AnimationNameProperty;
} & Omit<
  StandardProperties<TLength> &
    VendorProperties<TLength> &
    ObsoleteProperties<TLength> &
    SvgProperties<TLength>,
  | 'animationName'
  | 'fontFamily'
  | 'MozAnimationName'
  | 'WebkitAnimationName'
  | 'OAnimationName'
>;
