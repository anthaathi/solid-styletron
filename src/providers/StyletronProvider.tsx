import type { Component, JSX } from 'solid-js';
import type { StandardEngine } from 'styletron-standard';
import {
  StyletronContext,
  StyletronTheme,
  StyletronThemeContext,
} from '../contexts/styletron';

export interface StyletronProviderProps {
  client: StandardEngine;
  theme: StyletronTheme;
  children: JSX.Element;
}

export const StyletronProvider: Component<StyletronProviderProps> = (props) => {
  return (
    <StyletronContext.Provider value={props.client}>
      <StyletronThemeContext.Provider value={props.theme}>
        {props.children}
      </StyletronThemeContext.Provider>
    </StyletronContext.Provider>
  );
};
