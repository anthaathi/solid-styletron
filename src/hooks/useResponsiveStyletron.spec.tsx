import { useResponsiveStyletron } from './useResponsiveStyletron';
import { StyletronProvider } from '../providers/StyletronProvider';
import { Server } from 'styletron-engine-atomic';
import { cleanup, render } from 'solid-testing-library';

describe('useResponsiveStyletron', () => {
  afterEach(() => cleanup());

  it('should render the stuff', async () => {
    const client = new Server();

    const Component = () => {
      const [css] = useResponsiveStyletron();

      return (
        <div
          class={css({
            margin: ['12px'],
            padding: '12px',
          })}
        />
      );
    };

    const App = () => {
      return (
        <StyletronProvider
          client={client}
          theme={{
            mediaQuery: { sm: '@media screen and (max-width: 1200px)' },
          }}
        >
          <Component />
        </StyletronProvider>
      );
    };

    const ele = await render(() => (
      <div>
        <App />
        {client.getStylesheets().map((res) => {
          return <style media={res.attrs.media}>{res.css}</style>;
        })}
      </div>
    ));

    expect(ele.container).toMatchSnapshot();
  });
});
