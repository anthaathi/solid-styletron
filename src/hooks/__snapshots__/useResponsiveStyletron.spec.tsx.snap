// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useResponsiveStyletron should render the stuff 1`] = `
<div>
  <div>
    <div
      class="ae af"
    />
    <style>
      .af{padding:12px}
    </style>
    <style
      media="screen and (max-width: 1200px)"
    >
      .ae{margin:12px}
    </style>
  </div>
</div>
`;
